File: typescript/0100_SameTree.ts
// Problem 100: Same Tree
// https://leetcode.com/problems/same-tree/description/

// Definition for a binary tree node.
class TreeNode {
  val: number
  left: TreeNode | null
  right: TreeNode | null
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = (val===undefined ? 0 : val)
    this.left = (left===undefined ? null : left)
    this.right = (right===undefined ? null : right)
  }
}

function isSameTree(p: TreeNode | null, q: TreeNode | null): boolean {
  const df = function(r: TreeNode | null, s: TreeNode | null): boolean{
    // 如果兩個都到底了, 回傳true
      if (!r && !s) return true; 
      // 如果只有其中一個到底, 回傳false
      if (!r || !s) return false; 
      // 如果兩個的值相同, 就繼續比對這兩個的左邊 和 這兩個的右邊, 回傳 and 的結果(要左比左相同, 右比右相同才會是true)
      if (r.val === s.val) return (df(r.right, s.right) && df(r.left, s.left));
      // 其餘都是false
      return false;
  };
  return df(p, q);
};

File: java/0100_SameTree.java
// Problem 100: Same Tree
// https://leetcode.com/problems/same-tree/

// Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public boolean isSameTree(TreeNode p, TreeNode q) {
        return dfs(p, q);
    }

    private boolean dfs(TreeNode r, TreeNode s) {
        // 如果兩個都到底了, 回傳true
        if (r == null && s == null) return true;
        // 如果只有其中一個到底, 回傳false
        if (r == null || s == null) return false;
        // 如果兩個的值相同, 就繼續比對這兩個的左邊 和 這兩個的右邊, 回傳 and 的結果(要左比左相同, 右比右相同才會是true)
        if (r.val == s.val) return dfs(r.right, s.right) && dfs(r.left, s.left);
        // 其餘都是false
        return false;
    }
}